File: android/src/main/kotlin/com/yeeb/hypertrack_views_flutter/toMapExtensions.kt
package com.yeeb.hypertrack_views_flutter

import com.hypertrack.sdk.views.dao.MovementStatus
import com.hypertrack.sdk.views.dao.Trip

fun Trip.toMap(): HashMap<String, Any?> {
    return hashMapOf(
            "trip_id" to this.tripId,
            "status" to this.status,
            "startedAt" to this.startedAt,
            "metadata" to this.metadata,
            "summary" to this.summary,
            "destination.address" to this.destination?.address,
            "destination.radius" to this.destination?.radius,
            "destination.latitude" to this.destination?.latitude,
            "destination.longitude" to this.destination?.longitude,
            "destination.arrivedAt" to this.destination?.arrivedAt,
            "destination.exitedAt" to this.destination?.exitedAt,
            "destination.scheduledAt" to this.destination?.scheduledAt
    )
}

fun MovementStatus.toMap(): HashMap<String, Any?> {
    return hashMapOf(
            "device_id" to this.deviceId,

            "device_info.app_version_number" to  this.deviceInfo?.appVersionNumber,
            "device_info.app_version_string" to this.deviceInfo?.appVersionString,
            "device_info.device_brand" to this.deviceInfo?.deviceBrand,
            "device_info.device_model" to this.deviceInfo?.deviceModel,
            "device_info.name" to this.deviceInfo?.name,
            "device_info.os_name" to this.deviceInfo?.osName,
            "device_info.os_version" to this.deviceInfo?.osVersion,
            "device_info.sdk_version" to this.deviceInfo?.sdkVersion,
            "device_info.battery" to this.battery,

            "device_status.createdAt" to this.deviceStatus?.createdAt,
            "device_status.status" to this.deviceStatus?.status,

            "location.accuracy" to this.location?.accuracy,
            "location.altitude" to this.location?.altitude,
            "location.bearing" to this.location?.bearing,
            "location.speed" to this.location?.speed,
            "location.latitude" to this.location?.latitude,
            "location.longitude" to this.location?.longitude,
            "location.recordedAt" to this.location?.recordedAt,

            "trips" to this.trips.map { it.toMap() }
    )
}

File: ios/Classes/toMapExtensions.swift
//
//  toMapExtensions.swift
//  hypertrack_views_flutter
//
//  Created by Ethan on 1/21/21.
//

import Foundation
import HyperTrackViews

extension MovementStatus.Trip {
    func toMap() -> [String: Any?] {
        var arriveAt: Any?
        var polyline: Any?
        switch destination?.estimate {
        case let .relevant(route):
            arriveAt = route.duration.description
            polyline = route.polyline.map {
                (coordinate) -> [Double] in
                [coordinate.latitude, coordinate.longitude]
            }
        default:
            break
        }
        return [
            "trip_id": id,
            "startedAt": startedAt.description,
            "metadata": metadata,
            "summary": summary,
            "embedURL": views.embedURL.absoluteString,
            "shareURL": views.shareURL.absoluteString,

            "destination.address": destination?.address,
            "destination.radius": destination?.radius,
            "destination.latitude": destination?.coordinate.latitude,
            "destination.longitude": destination?.coordinate.longitude,
            "destination.arrivedAt": destination?.arrivedAt?.description,
            "destination.exitedAt": destination?.exitedAt?.description,
            "destination.scheduledAt": destination?.scheduledAt?.description,

            "estimate.arriveAt": arriveAt,
            "estimate.route": polyline,
        ]
    }
}

extension MovementStatus {
    func toMap() -> [String: Any?] {
        var batteryState: Int
        var statusState: Int
        switch battery {
        case .low:
            batteryState = 0
        case .normal:
            batteryState = 1
        case .charging:
            batteryState = 2
        default:
            batteryState = -1
        }
        switch status.type {
        case .active:
            statusState = 0
        case .disconnected:
            statusState = 1
        case .inactive:
            statusState = 2
        default:
            statusState = -1
        }
        var movementStatus: [String: Any?] = [
            "device_id": device.id,

            "device_info.name": device.name,
            "device_info.battery": batteryState,

            "device_status.createdAt": status.timestamp.description,
            "device_status.status": statusState,

            "location.accuracy": location.horizontalAccuracy,
            "location.altitude": location.altitude,
            "location.bearing": location.course,
            "location.speed": location.speed,
            "location.latitude": location.coordinate.latitude,
            "location.longitude": location.coordinate.longitude,
            "location.recordedAt": location.timestamp.description,

            "trips": trips.map { (trip) -> [String: Any?] in
                trip.toMap()
            },
        ]
        switch device.os {
        case let .Android(os):
            movementStatus["device_info.app_version_number"] = os.app.versionCode
            movementStatus["device_info.app_version_string"] = os.app.version
            movementStatus["device_info.device_brand"] = os.hardware.brand
            movementStatus["device_info.device_model"] = os.hardware.model
            movementStatus["device_info.os_name"] = "Android"
            movementStatus["device_info.os_version"] = os.version
            movementStatus["device_info.sdk_version"] = os.sdk.version
        case let .iOS(os):
            movementStatus["device_info.app_version_number"] = os.app.id
            movementStatus["device_info.app_version_string"] = os.app.version
            movementStatus["device_info.device_brand"] = "Apple"
            movementStatus["device_info.device_model"] = os.hardware.model
            movementStatus["device_info.os_name"] = "iOS"
            movementStatus["device_info.os_version"] = os.version
            movementStatus["device_info.sdk_version"] = os.sdk.version
        }

        return movementStatus
    }
}
